### PR TITLE
Consolidate time format for API

### LINKED ISSUE #18276: Remote API `/version` request `BuildTime` field weird date format
I'm maintaining the remote API client for Dart and wonder why there are different date/time formats used.
I have a `parseDate()` function where I parse a format like 
- `'0001-01-01T00:00:00Z'` (near ISO 8601 but not quite)
- milliseconds since epoche

and now in Docker 1.9 I again encounter a new format in the `/version` response `BuildTime` field `'Fri Nov 20 17:56:04 UTC 2015'` which I find expecially cumbersome to parse.

Please stick at least to the former two or switch everything to ISO 8601 or some other proper format.

### PR DESCRIPTION
**Pull request by WeiZhang555**
Fixes #18276

Consolidate all the API to same time format: RFC3339, and it will be
client's responsibility to present it in more user friendly way.

Signed-off-by: Zhang Wei zhangwei555@huawei.com

This is an initial commit for fixing the issue, I believe that there must be other time related struct elements need to be formatted, also am investigating the influence on other components...
I will update in following commit.

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks for creating a PR @WeiZhang555!

**Comment by calavera** [comment]
LGTM

**Comment by WeiZhang555** [comment]
@calavera Thank you for help reviewing! :smile: 

I just changed format from "time.RFC3339" to "time.RFC3339Nano", because I found that `time.Time.MarshalJSON` use this format as default and it's better to keep them exactly the same:

```
func (t Time) MarshalJSON() ([]byte, error)

MarshalJSON implements the json.Marshaler interface. The time is a quoted string in RFC 3339 format, with sub-second precision added if present.
```

As time related elements of all API structs are already of format RFC3339Nano, I think `docker version` is the only part need to change, also I didn't see bad influence introduced by this PR, so I think this is already all what we have to do (easier than I thought :smile: )

PTAL! Thank you all!

**Comment by thaJeztah** [comment]
Thanks @WeiZhang555! We're not in docs review yet, but I noticed the `BuildTime` field is not documented in the API documentation. Can you;

- Add the field to the example response in; https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.22.md#display-system-wide-information
- Add a note to the [API changelog](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v122-api-changes), e.g. `GET /info` now returns the `BuildTime` field in RFC3339Nano format to make it consistent with other date/time values returned by the API.

**Comment by vdemeester** [comment]
LGTM 🐯

**Comment by tiborvass** [comment]
I don't see any change in the API, it's only in the client-side. Am I missing something?

**Comment by thaJeztah** [comment]
@tiborvass I think the API displays the output that's generated in `hack/make.sh`, so no API changes in golang, but it will be a different output than previously

**Comment by tiborvass** [comment]
@thaJeztah i don't understand. I thought the issue was that the API doesn't return time in the standard format.

**Comment by thaJeztah** [comment]
@tiborvass it didn't, that's why `hack/make.sh` is modified in this PR; https://github.com/docker/docker/pull/18309/files#diff-c4b8d2fa793507a7ce5835e7ca780464R73

The other code changes in this PR are to make sure that the output on the CLI is still the same as before (human readable / friendly)

**Comment by tiborvass** [comment]
@thaJeztah ok my bad, didn't see this was just build time :)

LGTM

**Comment by thaJeztah** [comment]
For docs; see my earlier comment: https://github.com/docker/docker/pull/18309#issuecomment-160931462

**Comment by WeiZhang555** [comment]
@thaJeztah Docs have been updated following your guide, please take a look :)

File: docs/reference/api/docker_remote_api_v1.22.md [hunk 1]
```diff
@@ -1990,14 +1990,15 @@ Show the docker version information
...
-         "GitCommit": "a8a31ef",
+         "KernelVersion": "3.19.0-23-generic",
+         "GoVersion": "go1.4.2",
+         "GitCommit": "e75da4b",
          "Arch": "amd64",
-         "ApiVersion": "1.20",
-         "Experimental": false
+         "ApiVersion": "1.22",
+         "BuildTime": "2015-12-01T07:09:13.444803460+00:00",
+         "Experimental": true
```
Comment (thaJeztah) [inline docs/reference/api/docker_remote_api_v1.22.md]: hehe, most people probably won't see "Experimental", but I don't think it hurts :-)

**Comment by thaJeztah** [comment]
Thanks for updating the docs @WeiZhang555!

docs LGTM

ping @moxiegirl @SvenDowideit @jamtur01

File: docs/reference/api/docker_remote_api.md [hunk 2]
```diff
@@ -99,6 +99,8 @@ This section lists each version from latest to oldest.  Each listing includes a
 * `GET /info` Now returns `Architecture` and `OSType` fields, providing information
   about the host architecture and operating system type that the daemon runs on.
 * `GET /networks/(name)` now returns a `Name` field for each container attached to the network.
+* `GET /version` now returns the `BuildTime` field in RFC3339Nano format to make it 
+  consistent with other date/time values returned by the API.
```
Comment (moxiegirl) [inline docs/reference/api/docker_remote_api.md]: @WeiZhang555 Thank you!  Do you think we need to link to the format or will people know the format?  @thaJeztah ?

Comment (thaJeztah) [inline docs/reference/api/docker_remote_api.md]: well, we use the same format in all other places. might be worth describing it in general (e.g. "date/time fields in the API use the `RFC3339` standard", but that's something for another PR

**Comment by thaJeztah** [comment]
Thanks for another contribution, @WeiZhang555!

**Comment by jhowardmsft** [comment]
Actually, this has broken Windows CI. The version of bash on the Windows CI machines doesn't support rfc-3339 in it's formatting. Just iso-8601 or rfc-822. Here's the date --help output from one of the CI servers:
```
Jenkins@JENKINS-PRS-1 /c/jenkins/gopath/src/github.com/docker/docker (master)
$ date --help
Usage: date [OPTION]... [+FORMAT]
  or:  date [OPTION] [MMDDhhmm[[CC]YY][.ss]]
Display the current time in the given FORMAT, or set the system date.

  -d, --date=STRING         display time described by STRING, not `now'
  -f, --file=DATEFILE       like --date once for each line of DATEFILE
  -I, --iso-8601[=TIMESPEC] output an ISO-8601 compliant date/time string.
                            TIMESPEC=`date' (or missing) for date only,
                            `hours', `minutes', or `seconds' for date and
                            time to the indicated precision.
  -r, --reference=FILE      display the last modification time of FILE
  -R, --rfc-822             output RFC-822 compliant date string
  -s, --set=STRING          set time described by STRING
  -u, --utc, --universal    print or set Coordinated Universal Time
      --help                display this help and exit
      --version             output version information and exit

FORMAT controls the output.  The only valid option for the second form
specifies Coordinated Universal Time.  Interpreted sequences are:

  %%   a literal %
  %a   locale's abbreviated weekday name (Sun..Sat)
  %A   locale's full weekday name, variable length (Sunday..Saturday)
  %b   locale's abbreviated month name (Jan..Dec)
  %B   locale's full month name, variable length (January..December)
  %c   locale's date and time (Sat Nov 04 12:02:33 EST 1989)
  %d   day of month (01..31)
  %D   date (mm/dd/yy)
  %e   day of month, blank padded ( 1..31)
  %h   same as %b
  %H   hour (00..23)
  %I   hour (01..12)
  %j   day of year (001..366)
  %k   hour ( 0..23)
  %l   hour ( 1..12)
  %m   month (01..12)
  %M   minute (00..59)
  %n   a newline
  %p   locale's AM or PM
  %r   time, 12-hour (hh:mm:ss [AP]M)
  %s   seconds since 00:00:00, Jan 1, 1970 (a GNU extension)
  %S   second (00..60)
  %t   a horizontal tab
  %T   time, 24-hour (hh:mm:ss)
  %U   week number of year with Sunday as first day of week (00..53)
  %V   week number of year with Monday as first day of week (01..52)
  %w   day of week (0..6);  0 represents Sunday
  %W   week number of year with Monday as first day of week (00..53)
  %x   locale's date representation (mm/dd/yy)
  %X   locale's time representation (%H:%M:%S)
  %y   last two digits of year (00..99)
  %Y   year (1970...)
  %z   RFC-822 style numeric timezone (-0500) (a nonstandard extension)
  %Z   time zone (e.g., EDT), or nothing if no time zone is determinable

By default, date pads numeric fields with zeroes.  GNU date recognizes
the following modifiers between `%' and a numeric directive.

  `-' (hyphen) do not pad the field
  `_' (underscore) pad the field with spaces

Report bugs to <bug-sh-utils@gnu.org>.

Jenkins@JENKINS-PRS-1 /c/jenkins/gopath/src/github.com/docker/docker (master)
$
```

**Comment by WeiZhang555** [comment]
@jhowardmsft Oh no, I totally forgot Windows, actually I don't know how to test it on Windows because I don't have appropriate platform. Sorry for breaking Windows CI and thank you for helping it out! :laughing:

**Comment by jhowardmsft** [comment]
@WeiZhang555 Happy to help. Hopefully if I can get Windows CI back running and reliable again (I've been working on it for days now), these types of things will be caught.

**Pull request merged**